File: frontend/src/lib/models/page_elements/text_input_element_config.ts
import type { PageElementConfig } from "../page_element_config";
import type { Variable } from "../value_configs";

export let TEXT_INPUT_ELEMENT_TYPE = "textinput";

export interface TextInputElementConfig  extends PageElementConfig {
    controller: TextInputController;
    label?: string;
    placeholder?: Variable;
    options?: TextInputOptions;
}

export interface TextInputController {
    input_identifier: string;
    binding_variable: Variable;
    input_validation?: TextInputValidation[];
    required?: boolean;
}

export interface TextInputOptions {
    type?: string;
}

export let TEXT_INPUT_VALIDATION_TYPE_REGEX = "regex";
export let TEXT_INPUT_VALIDATION_TYPE_MIN_LENGTH = "min_length";
export let TEXT_INPUT_VALIDATION_TYPE_EQUAL_TO_INPUT = "eq_input";

export interface TextInputValidation {
    type: string;
    value: string;
}

File: src/main/kotlin/dev/schlaubi/dashboard/models/pageelements/TextInputElementConfig.kt
package dev.schlaubi.dashboard.models.pageelements

import dev.schlaubi.dashboard.models.PageVariable
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
@SerialName("textinput")
data class TextInputElementConfig(
    val controller: TextInputController,
    val label: String? = null,
    val placeholder: PageVariable? = null,
    val options: TextInputOptions? = null,
) : PageElementConfig()

@Serializable
data class TextInputController(
    @SerialName("input_identifier") val inputIdentifier: String,
    @SerialName("binding_variable") val bindingVariable: PageVariable,
    @SerialName("input_validation") val inputValidation: List<TextInputValidation>? = null,
    val required: Boolean = true,
)

@Serializable
data class TextInputOptions(val type: String? = null)

@Serializable
enum class TextInputValidationType {
    @SerialName("regex")
    REGEX,

    @SerialName("min_length")
    MIN_LENGTH,

    @SerialName("eq_input")
    EQ_INPUT
}

@Serializable
data class TextInputValidation(val type: TextInputValidationType, val value: String)
